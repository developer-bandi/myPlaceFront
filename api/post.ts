import {
  addCommentRes,
  addLikecountRes,
  addPostDetailRes,
  deleteCommentRes,
  deleteLikecountRes,
  deletePostDetailRes,
  getPostDetailRes,
  getPostListRes,
  getSearchPostListRes,
} from "../type/post";
import axiosInstance from "./core";
import makeForm from "./lib/makeForm";
interface postListCondition {
  page: number;
  order: string;
}
interface searchPostListCondition extends postListCondition {
  keyword: string;
}
interface addCommentInfo {
  PostId: number;
  content: string;
}
interface deleteCommentInfo {
  CommentId: number;
  UserId: number;
}
interface deletePostInfo {
  PostId: number;
  UserId: number;
}
interface addPostInfo {
  [key: string]: unknown;
  title: string;
  content: string;
  imgs: Blob[];
}

export const getPostList = ({ page, order }: postListCondition) =>
  axiosInstance.get<getPostListRes>(`post/list?page=${page}&order=${order}`);

export const getSearchPostList = ({
  keyword,
  page,
  order,
}: searchPostListCondition) =>
  axiosInstance.get<getSearchPostListRes>(
    `post/search?keyword=${keyword}&page=${page}&order=${order}`
  );

export const getPostDetail = (id: string) =>
  axiosInstance.get<getPostDetailRes>(`post/detail?id=${id}`);

export const addPostDetail = (info: addPostInfo) =>
  axiosInstance.post<addPostDetailRes>("post/detail", makeForm(info));

export const deletePostDetail = (postInfo: deletePostInfo) =>
  axiosInstance.delete<deletePostDetailRes>("post/detail", {
    data: postInfo,
  });

export const addComment = (commentInfo: addCommentInfo) =>
  axiosInstance.post<addCommentRes>("post/comment", commentInfo);

export const deleteComment = (commentInfo: deleteCommentInfo) =>
  axiosInstance.delete<deleteCommentRes>("post/comment", {
    data: commentInfo,
  });

export const addLikecount = (PostId: string) =>
  axiosInstance.post<addLikecountRes>("post/likecount", {
    PostId,
  });

export const deleteLikecount = (PostId: string) =>
  axiosInstance.delete<deleteLikecountRes>("post/likecount", {
    data: {
      PostId,
    },
  });
