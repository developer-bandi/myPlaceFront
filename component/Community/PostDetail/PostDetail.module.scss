.mainBlock {
  width: 1000px;
  margin: 50px auto 50px auto;
  min-height: calc(100vh - 282px);
  @media screen and (max-width: 1023px) {
    width: calc(100% - 40px);
    margin: 30px 20px 30px 20px;
    min-height: calc(100vh - 242px);
  }
}

.titleBlock {
  border-bottom: 1px solid #eaeaea;
  padding: 0 0 10px 0;
  width: 100%;
}

.title {
  font-size: 30px;
  font-weight: 500;
  margin: 0 0 10px 0;
  width: 100%;
}

.titleInfoBlock {
  display: flex;
}

.leftBlock {
  display: flex;
}

.nickname {
  color: #50575e;
  margin: 0 10px 0 0;
}

.date {
  color: #c5cec1;
}

.rightBlock {
  display: flex;
  margin: 0 0 0 auto;
}

.deletePostButton {
  width: 70px;
  margin: 0 0 0 10px;
  cursor: pointer;
}

.viewCount {
  display: flex;
  align-items: center;
  margin: 0 0 0 10px;
}

.content {
  padding: 30px 0 100px 0;
  line-height: 1.8;
}

.likeCount {
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid #e5e5e5;
  padding: 20px;
  width: 90px;
  height: 50px;
  border-radius: 5px;
  margin: 0 auto 100px auto;
}

.commentBlock {
  border-bottom: 1px solid #eaeaea;
  align-items: flex-start;
  padding: 10px;
}

.commentTitle {
  width: 1000px;
  display: flex;
  align-items: center;
  border-bottom: 1px solid #eaeaea;
  padding: 0 0 10px 0;
  @media screen and (max-width: 1023px) {
    width: 100%;
  }
}

.icon {
  margin: 0 5px 0 0;
}

.commentUser {
  font-size: 15px;
  font-weight: 500;
}

.commentContent {
  width: 100%;
  line-height: 1.8;
}

.commentDate {
  font-size: 15px;
  margin: 0 0 0 10px;
  text-align: right;
  color: #c5cec1;
}

.deleteCommentButton {
  margin: 0 0 0 10px;
  padding: 0;
  cursor: pointer;
}

.commentInputBlock {
  display: flex;
  align-items: center;
  flex-wrap: wrap;
  height: 140px;
  border: 2px solid #cecece;
  padding: 10px;
  margin: 10px 0 0 0;
}

.commentInput {
  width: 900px;
  height: 120px;
}

.commentSubmitButton {
  background-color: black;
  color: white;
  width: 65px;
  height: 30px;
  margin: 85px 0 0 10px;
  border-radius: 5px;
}

.commentListBlock {
  width: 1000px;
  @media screen and (max-width: 1023px) {
    width: 100%;
  }
}

.commentInfoBlock {
  display: flex;
  align-items: center;
  margin: 0 0 10px0;
}

.photoBlock {
  position: relative;
  max-width: 100%;
  height: auto;
  @media screen and (max-width: 1023px) {
    width: 100%;
  }
}

.img {
  max-width: 100%;
  height: auto;
  object-fit: contain;
}
