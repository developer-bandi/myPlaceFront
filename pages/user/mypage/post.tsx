import Head from "next/head";
import { useRouter } from "next/router";
import Footer from "../../../component/Common/Footer/Footer";
import HeaderContainer from "../../../component/Common/Header/HeaderContainer";
import MyPostContainer from "../../../component/User/Mypage/MyPost/MyPostContainer";
import { useLogin } from "../../../lib/customHook/loginCheck";

const Post = () => {
  useLogin();
  const router = useRouter();
  return (
    <>
      <Head>
        <title>My place - 내 포스트</title>
        <meta
          name="title"
          property="og:title"
          content="My place -  내 포스트"
        />
        <meta
          name="description"
          property="og:description"
          content="내가 쓴 포스트를 볼수 있습니다"
        />
        <meta
          name="url"
          property="og:url"
          content={process.env.NEXT_PUBLIC_SERVER_DOMAIN + router.pathname}
        />
      </Head>
      <HeaderContainer />
      <MyPostContainer />
      <Footer />
    </>
  );
};

export default Post;
