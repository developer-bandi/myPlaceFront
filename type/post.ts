import { pagenationList } from "./common";

interface user {
  id: number;
  nickname: string;
}

export interface post {
  id: number;
  title: string;
  content: string;
  viewCount: number;
  createdAt: string;
  nickname: string;
  comment: number;
  postlikecount: number;
}

interface comment {
  id: number;
  content: string;
  createdAt: string;
  User: user;
}

interface postDetail
  extends Omit<post, "comment" | "postlikecount" | "nickname"> {
  Comments: comment[];
  Photos: string[];
  likelist: number[];
  User: user;
}

export type getPostListRes = pagenationList<post>;

export type getSearchPostListRes = pagenationList<post>;

export type getPostDetailRes = postDetail;

export type addPostDetailRes = string;

export type deletePostDetailRes = string;

export type addCommentRes = comment;

export type deleteCommentRes = string;

export type addLikecountRes = string;

export type deleteLikecountRes = string;
