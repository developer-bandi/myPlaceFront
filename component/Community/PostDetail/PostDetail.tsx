import styles from "./PostDetail.module.scss";
import { BiCommentDetail } from "react-icons/bi";
import { FcLike } from "react-icons/fc";
import { GrView } from "react-icons/gr";
import { AiOutlineHeart } from "react-icons/ai";
import { RefObject } from "react";
import { signupState } from "../../../store/reducers/userLogin/userLoginReducer";
import { setDateLatest } from "../../../lib/commonFn/date";
import { postDetailType } from "../../../lib/apitype/post";

interface PostDetailProps {
  serverSideData: postDetailType;
  textareaRef: RefObject<HTMLTextAreaElement>;
  comments: {
    PostId: number;
    User: {
      id: number;
      nickname: string;
    };
    UserId: number;
    content: string;
    createdAt: string;
    id: number;
    updatedAt: string;
  }[];
  likelist: number[];
  postComment: () => Promise<void>;
  loginedUser: signupState;
  deleteOrPostLikecount: () => Promise<void>;
  deleteComment: (CommentId: number, UserId: number) => Promise<void>;
  deletePost: (PostId: number, UserId: number) => Promise<void>;
}

const PostDetail = ({
  serverSideData,
  textareaRef,
  comments,
  likelist,
  postComment,
  loginedUser,
  deleteOrPostLikecount,
  deleteComment,
  deletePost,
}: PostDetailProps) => {
  return (
    <main className={styles.mainBlock}>
      <h3 className={styles.hidden}>{serverSideData.title}</h3>
      <div className={styles.titleBlock}>
        <h3 className={styles.title}>{serverSideData.title}</h3>
        <div className={styles.titleInfoBlock}>
          <ul className={styles.leftBlock}>
            <li className={styles.nickname}>{serverSideData.User.nickname}</li>
            <li className={styles.date}>
              {setDateLatest(serverSideData.createdAt)}
            </li>
            <li className={styles.viewCount}>
              <GrView className={styles.icon} />
              {serverSideData.viewCount}
            </li>
          </ul>
          <div className={styles.rightBlock}>
            {loginedUser.userInfo.id == serverSideData.UserId ? (
              <button
                className={styles.deletePostButton}
                onClick={() =>
                  deletePost(serverSideData.id, serverSideData.UserId)
                }
              >
                삭제하기
              </button>
            ) : null}
          </div>
        </div>
      </div>
      <div className={styles.photoListBlock}>
        {serverSideData.Photos.map((srcObj: { filename: string }) => {
          return (
            <div className={styles.photoBlock}>
              <img
                src={`${process.env.NEXT_PUBLIC_SERVER_DOMAIN}/imgs/${srcObj.filename}`}
                className={styles.img}
              />
            </div>
          );
        })}
      </div>
      <p className={styles.content}>{serverSideData.content}</p>
      <button className={styles.likeCount} onClick={deleteOrPostLikecount}>
        {likelist.indexOf(loginedUser.userInfo.id || -1) !== -1 ? (
          <FcLike className={styles.icon} />
        ) : (
          <AiOutlineHeart className={styles.icon} />
        )}
        {likelist.length}
      </button>
      <div className={styles.commentTitle}>
        <BiCommentDetail className={styles.icon} />
        댓글 {comments.length}개
      </div>
      <div className={styles.commentListBlock}>
        {comments.map((comment) => {
          return (
            <div className={styles.commentBlock}>
              <ul className={styles.commentInfoBlock}>
                <li className={styles.commentUser}>{comment.User.nickname}</li>

                <li className={styles.commentDate}>
                  {setDateLatest(comment.createdAt)}
                </li>
                {loginedUser.userInfo.id == comment.UserId ? (
                  <button
                    className={styles.deleteCommentButton}
                    onClick={() => deleteComment(comment.id, comment.UserId)}
                  >
                    삭제
                  </button>
                ) : null}
              </ul>
              <p className={styles.commentContent}>{comment.content}</p>
            </div>
          );
        })}
      </div>
      <div className={styles.commentInputBlock}>
        <textarea className={styles.commentInput} ref={textareaRef} />
        <button className={styles.commentSubmitButton} onClick={postComment}>
          등록
        </button>
      </div>
    </main>
  );
};

export default PostDetail;
