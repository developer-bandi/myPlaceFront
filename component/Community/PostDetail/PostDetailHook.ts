import {useRouter} from "next/router";
import {useRef, useState} from "react";
import {useSelector} from "react-redux";
import {postDetailType} from "../../../lib/apitype/post";
import {
  axiosDeleteComment,
  axiosDeletelikecount,
  axiosDeletePostDetail,
  axiosPostComment,
  axiosPostlikecount,
} from "../../../lib/commonFn/api";
import {RootReducer} from "../../../store";

const usePostDetail = (serverSideData: postDetailType) => {
  const loginedUser = useSelector((state: RootReducer) => state.userLogin);
  const [comments, setComments] = useState(serverSideData.Comments);
  const [likelist, setLikelist] = useState(serverSideData.likelist);
  const router = useRouter();
  const textareaRef = useRef<HTMLTextAreaElement>(null);

  const postComment = async () => {
    if (textareaRef.current !== null && loginedUser.content) {
      try {
        const newComment = await axiosPostComment(
          serverSideData.id,
          textareaRef.current.value
        );
        newComment.data.UserId = loginedUser.content.id;
        textareaRef.current.value = "";
        setComments([...comments, newComment.data]);
      } catch (error) {
        alert("에러가 발생하였습니다 다시 시도해주세요");
      }
    } else {
      alert("로그인 해주세요");
      router.push("/user/auth/signin");
    }
  };

  const deleteComment = async (CommentId: number, UserId: number) => {
    const agree = window.confirm("댓글을 삭제하시겠습니까?");
    if (agree) {
      try {
        await axiosDeleteComment(CommentId, UserId);
        setComments(
          comments.filter((data) => {
            if (data.id == CommentId) {
              return false;
            }
            return true;
          })
        );
        alert("댓글을 삭제하였습니다");
      } catch (error) {
        alert("에러가 발생했습니다 다시 시도해보세요");
      }
    }
  };

  const deletePost = async (PostId: number, UserId: number) => {
    const agree = window.confirm("포스트를 삭제하시겠습니까?");
    if (agree) {
      try {
        await axiosDeletePostDetail(PostId, UserId);
        router.push("/community/postlist");
      } catch (error) {
        alert("에러가 발생했습니다 다시 시도해보세요");
      }
    }
  };

  const deleteOrPostLikecount = async () => {
    if (loginedUser.content !== undefined) {
      try {
        if (likelist.indexOf(loginedUser.content.id) === -1) {
          await axiosPostlikecount(serverSideData.id);
          setLikelist([...likelist, loginedUser.content.id]);
        } else {
          await axiosDeletelikecount(serverSideData.id);
          const loginedId = loginedUser.content.id;
          setLikelist(
            likelist.filter((id: number) => {
              if (id == loginedId) {
                return false;
              }
              return true;
            })
          );
        }
      } catch (error) {
        alert("에러가 발생했습니다 다시 시도해보세요");
      }
    } else {
      alert("로그인을 진행해주세요");
    }
  };

  return {
    textareaRef,
    comments,
    likelist,
    loginedUser,
    postComment,
    deleteOrPostLikecount,
    deleteComment,
    deletePost,
  };
};

export default usePostDetail;
