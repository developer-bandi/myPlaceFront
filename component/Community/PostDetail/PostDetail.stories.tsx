import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import PostDetail from "./PostDetail";
import { INITIAL_VIEWPORTS } from "@storybook/addon-viewport";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

export default {
  title: "Community/PostDetail",
  component: PostDetail,
  parameters: {
    viewport: {
      viewports: INITIAL_VIEWPORTS,
      defaultViewport: "desktop",
    },
  },
} as ComponentMeta<typeof PostDetail>;

const Template: ComponentStory<typeof PostDetail> = () => (
  <Provider
    store={configureMockStore()({
      userLogin: {
        content: { id: 1, nickname: "test" },
        loading: false,
        error: false,
      },
      postDetail: {
        content: {
          Comments: [
            {
              PostId: 1,
              User: {
                id: 1,
                nickname: "test",
              },
              UserId: 1,
              content: "test댓글",
              createdAt: new Date(),
              id: 1,
              updatedAt: new Date(),
            },
          ],
          User: { id: 1, nickname: "test" },
          UserId: 1,
          content: "테스트용 컨텐츠",
          createdAt: new Date("2022-11-11").toDateString(),
          id: "1",
          likelist: [1, 2, 3],
          title: "테스트용 제목",
          updatedAt: new Date("2022-11-11").toDateString(),
          viewCount: 100,
          Photos: [{ filename: "aaaaa" }],
        },
      },
    })}
  >
    <PostDetail />
  </Provider>
);

export const generalType = Template.bind({});
