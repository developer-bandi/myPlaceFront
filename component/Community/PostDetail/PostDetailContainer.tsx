import { useRouter } from "next/router";
import { useRef, useState } from "react";
import { useSelector } from "react-redux";
import { postDetailType } from "../../../lib/apitype/post";
import {
  axiosDeleteComment,
  axiosDeletelikecount,
  axiosDeletePostDetail,
  axiosPostComment,
  axiosPostlikecount,
} from "../../../lib/commonFn/api";
import { RootReducer } from "../../../store";
import PostDetail from "./PostDetail";

interface PostDetailContainerProps {
  serverSideData: postDetailType;
}

const PostDetailContainer = ({ serverSideData }: PostDetailContainerProps) => {
  const loginedUser = useSelector((state: RootReducer) => state.userLogin);
  const textareaRef = useRef<HTMLTextAreaElement>(null);
  const [comments, setComments] = useState(serverSideData.Comments);
  const [likelist, setLikelist] = useState(serverSideData.likelist);
  const router = useRouter();

  const postComment = async () => {
    if (textareaRef.current !== null && loginedUser.loginStatus) {
      try {
        const newComment = await axiosPostComment(
          serverSideData.id,
          textareaRef.current.value
        );
        newComment.data.UserId = loginedUser.userInfo.id;
        setComments([...comments, newComment.data]);
        textareaRef.current.value = "";
      } catch (error) {
        alert("에러가 발생하였습니다 다시 시도해주세요");
      }
    } else {
      alert("로그인 해주세요");
      router.push("/user/auth/signin");
    }
  };

  const deleteComment = async (CommentId: number, UserId: number) => {
    const agree = window.confirm("댓글을 삭제하시겠습니까?");
    if (agree) {
      try {
        await axiosDeleteComment(CommentId, UserId);
        setComments(
          comments.filter((data) => {
            if (data.id == CommentId) {
              return false;
            }
            return true;
          })
        );
      } catch (error) {
        alert("에러가 발생했습니다 다시 시도해보세요");
      }
    }
  };

  const deletePost = async (PostId: number, UserId: number) => {
    const agree = window.confirm("포스트를 삭제하시겠습니까?");
    if (agree) {
      try {
        await axiosDeletePostDetail(PostId, UserId);
        router.push("/community/postlist");
      } catch (error) {
        alert("에러가 발생했습니다 다시 시도해보세요");
      }
    }
  };

  const deleteOrPostLikecount = async () => {
    if (loginedUser.loginStatus && loginedUser.userInfo.id !== undefined) {
      try {
        if (likelist.indexOf(loginedUser.userInfo.id) === -1) {
          await axiosPostlikecount(serverSideData.id);
          setLikelist([...likelist, loginedUser.userInfo.id]);
        } else {
          await axiosDeletelikecount(serverSideData.id);
          setLikelist(
            likelist.filter((id: number) => {
              if (id == loginedUser.userInfo.id) {
                return false;
              }
              return true;
            })
          );
        }
      } catch (error) {
        alert("에러가 발생했습니다 다시 시도해보세요");
      }
    } else {
      alert("로그인을 진행해주세요");
    }
  };

  return (
    <PostDetail
      serverSideData={serverSideData}
      textareaRef={textareaRef}
      comments={comments}
      likelist={likelist}
      loginedUser={loginedUser}
      postComment={postComment}
      deleteOrPostLikecount={deleteOrPostLikecount}
      deleteComment={deleteComment}
      deletePost={deletePost}
    />
  );
};

export default PostDetailContainer;
