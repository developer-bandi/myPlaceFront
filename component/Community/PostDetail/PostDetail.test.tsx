import {
  act,
  fireEvent,
  render,
  renderHook,
  screen,
} from "@testing-library/react";
import axios from "axios";
import {useRouter} from "next/router";
import React, {ReactNode} from "react";
import {Provider} from "react-redux";
import configureMockStore from "redux-mock-store";
import {postDetailType} from "../../../lib/apitype/post";
import PostDetail from "./PostDetail";
import usePostDetail from "./PostDetailHook";

jest.mock("axios");
jest.mock("next/router", () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

const axiosMock = axios as jest.Mocked<typeof axios>;
const postDetailMock: postDetailType = {
  Comments: [],
  Photos: [],
  User: {id: 1, nickname: "test"},
  UserId: 1,
  content: "test",
  createdAt: "test",
  id: 1,
  likelist: [],
  title: "test",
  updatedAt: "test",
  viewCount: 1,
};

describe("PostDetail Presentational 테스트", () => {
  const deleteOrPostLikecountMock = jest.fn();
  const deleteCommentMock = jest.fn();
  const deletePostMock = jest.fn();
  const postCommentMock = jest.fn();
  const utils = render(
    <PostDetail
      serverSideData={postDetailMock}
      textareaRef={{current: null}}
      comments={[
        {
          PostId: 1,
          User: {
            id: 1,
            nickname: "test",
          },
          UserId: 1,
          content: "test",
          createdAt: "test",
          id: 1,
          updatedAt: "test",
        },
      ]}
      likelist={[]}
      postComment={postCommentMock}
      loginedUser={{
        content: {id: 1, nickname: "test"},
        loading: false,
        error: false,
      }}
      deleteOrPostLikecount={deleteOrPostLikecountMock}
      deleteComment={deleteCommentMock}
      deletePost={deletePostMock}
    />
  );
  expect(utils.container).toMatchSnapshot();
  fireEvent.click(screen.getByTestId("deleteOrPostLikecount"));
  fireEvent.click(screen.getByTestId("deleteComment"));
  fireEvent.click(screen.getByTestId("deletePost"));
  fireEvent.click(screen.getByTestId("postComment"));

  expect(deleteOrPostLikecountMock).toBeCalledTimes(1);
  expect(deleteCommentMock).toBeCalledTimes(1);
  expect(deletePostMock).toBeCalledTimes(1);
  expect(postCommentMock).toBeCalledTimes(1);
});

describe("postDetail Hook 테스트", () => {
  let loginStatus = true;
  const mockStore = configureMockStore();
  const loginMockStore = mockStore({
    userLogin: {
      content: {id: 1, nickname: "test"},
      loading: false,
      error: false,
    },
  });
  const logoutMockStore = configureMockStore()({
    userLogin: {
      loading: false,
      error: false,
    },
  });
  const wrapper = ({children}: {children: ReactNode}) => (
    <Provider store={loginStatus ? loginMockStore : logoutMockStore}>
      {children}
    </Provider>
  );
  describe("postComment 테스트", () => {
    it("로그인 하여 댓글을 정상등록한 경우", async () => {
      const useRefSpy = jest
        .spyOn(React, "useRef")
        .mockReturnValue({current: {value: "test"}});
      axiosMock.post.mockImplementation(() =>
        Promise.resolve({
          status: 200,
          data: {
            PostId: 1,
            User: {
              id: 1,
              nickname: "test",
            },
            UserId: 1,
            content: "test",
            createdAt: "test",
            id: 1,
            updatedAt: "test",
          },
        })
      );
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      await act(async () => {
        await result.current.postComment();
      });

      expect(result.current.comments).toStrictEqual([
        {
          PostId: 1,
          User: {
            id: 1,
            nickname: "test",
          },
          UserId: 1,
          content: "test",
          createdAt: "test",
          id: 1,
          updatedAt: "test",
        },
      ]);
    });

    it("로그인 하였지만 에러가 발생한 경우", async () => {
      const useRefSpy = jest
        .spyOn(React, "useRef")
        .mockReturnValueOnce({current: {value: "test"}});
      window.alert = jest.fn();
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      axiosMock.post.mockImplementation(() =>
        Promise.reject({status: 500, data: "에러발생"})
      );

      await act(async () => {
        await result.current.postComment();
      });
      expect(window.alert).toBeCalledWith(
        "에러가 발생하였습니다 다시 시도해주세요"
      );
    });

    it("로그인 하지 않은 경우", async () => {
      loginStatus = false;
      const mockRouter = {
        push: jest.fn(),
      };
      (useRouter as jest.Mock).mockReturnValue(mockRouter);
      const useRefSpy = jest
        .spyOn(React, "useRef")
        .mockReturnValueOnce({current: "test"});
      window.alert = jest.fn();
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      await act(async () => {
        await result.current.postComment();
      });
      expect(window.alert).toBeCalledWith("로그인 해주세요");
      expect(mockRouter.push).toBeCalledWith("/user/auth/signin");
    });
  });

  describe("deleteComment 함수 테스트", () => {
    window.confirm = jest.fn(() => true);
    it("정상 삭제", async () => {
      postDetailMock.Comments.push({
        PostId: 1,
        User: {
          id: 1,
          nickname: "test",
        },
        UserId: 1,
        content: "test",
        createdAt: "test",
        id: 1,
        updatedAt: "test",
      });
      axiosMock.delete.mockImplementation(() =>
        Promise.resolve({
          status: 200,
          data: "test",
        })
      );
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      await act(async () => {
        await result.current.deleteComment(1, 1);
      });
      expect(window.alert).toBeCalledWith("댓글을 삭제하였습니다");
      expect(result.current.comments.length).toBe(0);
    });
    it("오류 발생", async () => {
      window.alert = jest.fn();
      axiosMock.delete.mockImplementation(() =>
        Promise.reject({status: 500, data: "에러발생"})
      );
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      await act(async () => {
        await result.current.deleteComment(1, 1);
      });
      expect(window.alert).toBeCalledWith(
        "에러가 발생했습니다 다시 시도해보세요"
      );
    });
  });
  describe("deletePost 함수 테스트", () => {
    window.confirm = jest.fn(() => true);
    it("정상 삭제", async () => {
      const mockRouter = {
        push: jest.fn(),
      };
      (useRouter as jest.Mock).mockReturnValue(mockRouter);
      axiosMock.delete.mockImplementation(() =>
        Promise.resolve({
          status: 200,
          data: "test",
        })
      );
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      await act(async () => {
        await result.current.deletePost(1, 1);
      });
      expect(mockRouter.push).toBeCalledWith("/community/postlist");
    });
    it("오류 발생", async () => {
      window.alert = jest.fn();
      axiosMock.delete.mockImplementation(() =>
        Promise.reject({status: 500, data: "에러발생"})
      );
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      await act(async () => {
        await result.current.deletePost(1, 1);
      });
      expect(window.alert).toBeCalledWith(
        "에러가 발생했습니다 다시 시도해보세요"
      );
    });
  });
  describe("deleteOrPostLikecount 함수 테스트", () => {
    it("좋아요 등록", async () => {
      loginStatus = true;
      axiosMock.post.mockImplementation(() =>
        Promise.resolve({
          status: 200,
          data: "test",
        })
      );
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      await act(async () => {
        await result.current.deleteOrPostLikecount();
      });
      expect(result.current.likelist).toEqual([1]);
    });

    it("좋아요 취소", async () => {
      axiosMock.delete.mockImplementation(() =>
        Promise.resolve({
          status: 200,
          data: "test",
        })
      );
      postDetailMock.likelist.push(1);
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      await act(async () => {
        await result.current.deleteOrPostLikecount();
      });
      expect(result.current.likelist).toEqual([]);
    });

    it("오류 발생", async () => {
      window.alert = jest.fn();
      axiosMock.delete.mockImplementation(() =>
        Promise.reject({status: 500, data: "에러발생"})
      );
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      await act(async () => {
        await result.current.deleteOrPostLikecount();
      });
      expect(window.alert).toBeCalledWith(
        "에러가 발생했습니다 다시 시도해보세요"
      );
    });

    it("로그인 하지 않음", async () => {
      loginStatus = false;
      window.alert = jest.fn();
      const {result} = renderHook(() => usePostDetail(postDetailMock), {
        wrapper,
      });
      await act(async () => {
        await result.current.deleteOrPostLikecount();
      });
      expect(window.alert).toBeCalledWith("로그인을 진행해주세요");
    });
  });
});
