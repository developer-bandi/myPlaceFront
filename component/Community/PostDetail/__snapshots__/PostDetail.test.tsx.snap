// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <main
    class="mainBlock"
  >
    <div
      class="titleBlock"
    >
      <h3
        class="title"
      >
        test
      </h3>
      <div
        class="titleInfoBlock"
      >
        <ul
          class="leftBlock"
        >
          <li
            class="nickname"
          >
            test
          </li>
          <li
            class="date"
          >
            NaN년 전
          </li>
          <li
            class="viewCount"
          >
            <svg
              class="icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12,21 C7,21 1,16 1,12 C1,8 7,3 12,3 C17,3 23,8 23,12 C23,16 17,21 12,21 Z M12,7 C9.23875,7 7,9.23875 7,12 C7,14.76125 9.23875,17 12,17 C14.76125,17 17,14.76125 17,12 C17,9.23875 14.76125,7 12,7 L12,7 Z"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
            1
          </li>
        </ul>
        <div
          class="rightBlock"
        >
          <button
            class="deletePostButton"
            data-testid="deletePost"
          >
            삭제하기
          </button>
        </div>
      </div>
    </div>
    <div
      class="photoListBlock"
    />
    <p
      class="content"
    >
      test
    </p>
    <button
      class="likeCount"
      data-testid="deleteOrPostLikecount"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 1024 1024"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M923 283.6a260.04 260.04 0 0 0-56.9-82.8 264.4 264.4 0 0 0-84-55.5A265.34 265.34 0 0 0 679.7 125c-49.3 0-97.4 13.5-139.2 39-10 6.1-19.5 12.8-28.5 20.1-9-7.3-18.5-14-28.5-20.1-41.8-25.5-89.9-39-139.2-39-35.5 0-69.9 6.8-102.4 20.3-31.4 13-59.7 31.7-84 55.5a258.44 258.44 0 0 0-56.9 82.8c-13.9 32.3-21 66.6-21 101.9 0 33.3 6.8 68 20.3 103.3 11.3 29.5 27.5 60.1 48.2 91 32.8 48.9 77.9 99.9 133.9 151.6 92.8 85.7 184.7 144.9 188.6 147.3l23.7 15.2c10.5 6.7 24 6.7 34.5 0l23.7-15.2c3.9-2.5 95.7-61.6 188.6-147.3 56-51.7 101.1-102.7 133.9-151.6 20.7-30.9 37-61.5 48.2-91 13.5-35.3 20.3-70 20.3-103.3.1-35.3-7-69.6-20.9-101.9zM512 814.8S156 586.7 156 385.5C156 283.6 240.3 201 344.3 201c73.1 0 136.5 40.8 167.7 100.4C543.2 241.8 606.6 201 679.7 201c104 0 188.3 82.6 188.3 184.5 0 201.2-356 429.3-356 429.3z"
        />
      </svg>
      0
    </button>
    <div
      class="commentTitle"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7 7h10v2H7zm0 4h7v2H7z"
        />
        <path
          d="M20 2H4c-1.103 0-2 .897-2 2v18l5.333-4H20c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zm0 14H6.667L4 18V4h16v12z"
        />
      </svg>
      댓글 
      1
      개
    </div>
    <div
      class="commentListBlock"
    >
      <div
        class="commentBlock"
      >
        <ul
          class="commentInfoBlock"
        >
          <li
            class="commentUser"
          >
            test
          </li>
          <li
            class="commentDate"
          >
            NaN년 전
          </li>
          <button
            class="deleteCommentButton"
            data-testid="deleteComment"
          >
            삭제
          </button>
        </ul>
        <p
          class="commentContent"
        >
          test
        </p>
      </div>
    </div>
    <div
      class="commentInputBlock"
    >
      <textarea
        aria-label="commentTextarea"
        class="commentInput"
      />
      <button
        class="commentSubmitButton"
        data-testid="postComment"
      >
        등록
      </button>
    </div>
  </main>
</div>
`;
